text/0000-named-arguments.md

```diff
@@ -0,0 +1,623 @@
+
+### Parameter hints in IDEs
+
+IDEs can provide hints for function arguments. Both IntelliJ-Rust (the plugin that provides Rust language support in JetBrains IDEs) and rust-analyzer (a Rust language server that is going to replace RLS) can do this. Many Rust programmers use these IDE-powered parameter hints, which proves that there is demand for named arguments, at least when reading code.
+
+Unfortunately, parameter hints are not available when reading code in a blog post, in an online forum, on GitHub or any other website. Furthermore, some editors aren't able to show parameter hints.
+
+Another problem is that IDEs are not perfect. For example, when reading code in a `#[cfg(target_os = "windows")]` module on a Linux machine, neither IntelliJ-Rust nor rust-analyzer can provide parameter hints.
+
+Therefore, I believe that Rust code shouldn't rely on an IDE to be readable and expressive.
+
+### Comments
+
+For example:
+
+```rust
+free_fall(/*z0*/ 100.0, /*v0*/ 0.0, /*g*/ 9.81);
+
+free_fall(100.0, 0.0, 9.81); // z0, v0, g
+
+free_fall(
+    100.0,  // z0
+    0.0,    // v0
+    9.81,   // g
+);
+```
+
+Since Rust code like this appears very rarely in the wild, it appears that most Rust programmers either aren't aware of this possibility, or don't like writing code like this.
+
+### Enums for `bool`
+
+Instead of a `bool` argument, a custom enum with descriptive variant names can be created, for example:
+
+```rust
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum CaseSensitivity {
+    CaseSensitive,
+    CaseInsensitive,
+}
+```
+
+This improves type safety, but it also has disadvantages:
+
+ * It adds boilerplate for the library author
+ * The custom enum must be `use`d at every call site, which also adds boilerplate for the library user
+ * It doesn't implement every trait `bool` implements
+ * Using it is more verbose. For example, `if is_case_sensitive {}` might become `if case_sensitivity == CaseSensitivity::CaseSensitive {}`
+
+### The newtype pattern
+
+The newtype pattern can be used to wrap arguments in structs with descriptive names:
+
+```rust
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Default)]
+pub struct Width(pub i32);
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Default)]
+pub struct Height(pub i32);
+
+pub fn set_size(Width(w): Width, Height(h): Height) {}
+```
+
+To reduce boilerplate, multiple values can be combined in one struct:
+
+```rust
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Default)]
+pub struct Size {
+    pub width: i32,
+    pub height: i32,
+}
+
+pub fn set_size(Size { width, height }: Size) {}
+```
+
+This also improves type safety, since a `Size` can't be passed to a function expecting a `Point`. However, this kind of type safety is often not required and not worth the additional complexity and boilerplate.
```

Comment:
This example is problematic, as replacing `Size` with `Rect` makes the example very reasonable. The width cannot exist without the height. I'm also confused why a codebase that deals with `Point` and `Rect` would not benefit from the additional type safety, while such errors are indeed quite common in codebases with geometric shapes and would not be fixed by the proposal.

Using the type as the differentiator between two otherwise structurally similar shapes is indeed good use.